Documentation/Relationship/Relationship.md

```diff
@@ -0,0 +1,598 @@
+to manually define or modify relationship information.
+
+**GraphQL:**  
+For the above example a GraphQL query could look like the following:
+```graphql
+query {
+  user(id: "532") {
+    name
+    lastname
+    post(id: "123") {
+      title
+    }
+    static-information {
+      info0
+      detailed {
+        detailedInfo0
+      }
+    }
+  }
+}
+```
+
+## 2. Manually defining Relationship Information
+
+This chapter imagines ways how a user could manually override or make modifications to relationship information.
+
+To provide such functionality the proposal introduces multiple `PathComponentModifiers`, similar to the already
+existing `ComponentModifiers`. As the `PathComponents` are passed as arguments to a `Group` instance,
+this could quickly get messy in terms of code readability.
+
+Thus this proposal additionally introduces a `PathComponentBuilder` with a corresponding initializer for the `Group` struct:
+
+```swift
+Group {
+  "path0".modifier0(arg0)
+  "path1"
+    .modifier1(arg1)
+    .modifier2(arg2)
+  $parameter
+    .modifier4(arg4)
+} content: {
+  Handler()
+}
+```
+
+### 2.1. Overriding the relationship name
+
+The in chapter [1.](#1-using-the-dsl-structure) described inference approach uses the string 
+description of the `PathComponent` as the relationship name. This may not fit everybody's needs.
+
+Such a customization mechanism is provided with the `.relationship` modifier (with the external parameter `name`):
+
+```swift
+var content: some Component {
+  Group {
+    "test".relationship(name: "new-name")
+  } content: {
+    Handler()
+  }
+}
+```
+
+The same `PathComponent` name cannot be used multiple times on the same depth in the `PathComponent` tree.
+Similarly the renamed Relationship MUST NOT collide with any other relationship name on the same depth.
+
+### 2.2. Adding new relationships
+
+To support rearranging components without breaking HATEOAS linking information or the GraphQL query schema,
+a user can also **add** their own relationship definitions.
+
+As we don't want users to manually list every `PathComponent` for the destination of the relationship, the user can create
+instances of `Relationship` which are then referenced with `.relatable` (marking destination) and
+`.relationship(to:)` modifiers.
+
+Those `Relationship` instances can be defined anywhere, preferably as a property of a `Component`.  
+Elsewise `Relationship` instances can also be organized into `RelationshipContainer` which are collected through
+the `relationshipContainer` property of a `Component`. Above modifiers then also accept KeyPaths into such containers.
+
+The example below showcases both ways of using `Relationship` instances:
+```swift
+struct TestServiceRelationShips: RelationshipContainer {
+    var greeter = Relationship("greeter")
+}
+
+struct TestService: WebService {
+
+  @PathParameter var userId: User.ID
+
+  var relationshipContainer = TestServiceRelationShips()
```

Comment:
The idea was that the `var relationshipContainer` is part of the `Component` protocol. Thus more or less the reason for the `RelationshipContainer` protocol. When visiting the `Component` one could collect all the different `RelationshipContainer` and saving their type to be able to match up `KeyPath`s to the different containers.
It would act the same way as a component modifier stored in the Context with scope `.environment`.
It was my approach trying to create some sort of key value storage for `Relationship` and making it easier to reference them.
But I agree that it probably is the least polished thing in this proposal and that there might be cleaner ways of doing this. As this is really a "syntactic sugar" kind of feature, I would suggest, that we leave it out of the proposal for now (at least if nobody has a better idea). I think there is currently a lot of infrastructure work ongoing which we might be able to use(?)